skimage/feature/_canny.py

```diff
@@ -14,44 +14,175 @@
 
 import numpy as np
 import scipy.ndimage as ndi
-from scipy.ndimage import generate_binary_structure, binary_erosion, label
+
 from ..filters import gaussian
-from .. import dtype_limits, img_as_float
+from .. import dtype_limits
 from .._shared.utils import check_nD
 
 
-def smooth_with_function_and_mask(image, function, mask):
-    """Smooth an image with a linear function, ignoring masked pixels.
+def _preprocess(image, mask, sigma, mode):
+    """Preprocess the image and mask before applying canny edge detection.
+
+    The image is smoothed using a gaussian filter ignoring masked
+    pixels and the mask is eroded.
 
     Parameters
     ----------
     image : array
-        Image you want to smooth.
-    function : callable
-        A function that does image smoothing.
+        Image to be smoothed.
     mask : array
         Mask with 1's for significant pixels, 0's for masked pixels.
+    sigma : scalar or sequence of scalars, optional
+        Standard deviation for Gaussian kernel. The standard
+        deviations of the Gaussian filter are given for each axis as a
+        sequence, or as a single number, in which case it is equal for
+        all axes.
+    mode : {'reflect', 'constant', 'nearest', 'mirror', 'wrap'}, optional
```

Comment:
Looks like this function is used only below, with mode='constant'. Is it useful to keep this option? I think nobody will touch it.
Otherwise, optional must be removed.

Reply:
In the previous implementation, `mode` is set to 'constant', but I think that the default recommended value 'nearest' used in `gaussian` must be used here also.
But this may be addressed in another PR ;)

Reply:
:coffee:

Reply:
+1 to check that in another PR.